src/system/BUILD.gn

```diff
@@ -212,14 +212,17 @@ source_set("system_config_header") {
   }
 }
 
+source_set("system_clock") {
+  sources = [ "SystemClock.h" ]
```

Comment:
Why only the header and not the CPP? This splits the dependency in a non-obvious way

Reply:
+1 - please never split h and cpp. I am working on untangling such splits from src/app and it is a nightmare. Let's not introduce this particular kind of code debt anymore as it is very hard to undo.

Reply:
Agree entirely. @anush-apple ?

Reply:
Ah, that was what I did first. I am not a ninja expert and would like some help solving this. Here is what happens when I tried to moved source to the same source set

```
diff --git a/src/system/BUILD.gn b/src/system/BUILD.gn
index 02bc73733b..b3383feaec 100644
--- a/src/system/BUILD.gn
+++ b/src/system/BUILD.gn
@@ -213,7 +213,10 @@ source_set("system_config_header") {
 }
 
 source_set("system_clock") {
-  sources = [ "SystemClock.h" ]
+  sources = [
+    "SystemClock.h",
+    "SystemClock.cpp",
+  ]
 }
 
 static_library("system") {
@@ -222,7 +225,6 @@ static_library("system") {
   sources = [
     "PlatformEventSupport.h",
     "SystemAlignSize.h",
-    "SystemClock.cpp",
     "SystemError.cpp",
     "SystemError.h",
     "SystemEvent.h",

```


Leads to this error:

../../../../../../../../../../src/system/SystemClock.cpp:25:10: fatal error: 'system/SystemClock.h' file not found
#include <system/SystemClock.h>
         ^~~~~~~~~~~~~~~~~~~~~~
../../../../../../../../../../src/system/SystemClock.cpp:25:10: note: did not find header 'SystemClock.h' in framework 'system' 

What is the right way to fix this @andy31415 @tcarmelveilleux ?

Reply:
Overall you have to look at all includes and ensure they are there.
In particular SystemClock.h depends on:
- SystemConfig.h => this means a "system_config_header" dependency
- lib/support/* => this means lib/support dependency
- system/SystemError => this means dependency on whatever contains this. You may need to split this out as well then since you are pulling out SystemClock and you cannot have reverse/circular dependencies

Then SystemClock.cpp depends on similar things so you are good with the above. The `SystemError.h` seems to be  the most painful one.

Reply:
Alternatives would be to make tracing depend on system ... however since tracing is so low level, it may not be possible (we may want to trace system itself and then we cannot be circular).

This may be another argument for why we cannot add tracing to all existing macros - verifyorexit for example is used in very low level code.

Reply:
General design of tracing is that:
  - tracing does not depend on much (at least not at the macro levels)
  - types for tracing are just "forward declare" and if something needs the actual thing, they need to import the correct header (e.g. a backed impl)

The difficulty you have is that you try to make the type strongly-typed visible everywhere. If that is the requirement (sure sounds like it) then we need to split out clock.

Reply:
Why do we have a timestamp if the backend is capable of calling `Time::now` by itself? what is the use case here (since I have been told RAII scopes is not what we are after)?

Reply:
see https://github.com/project-chip/connectedhomeip/pull/32223/files#r1497791502

just drop the timePoint. Backends can call `now()` when they receive the event. We do not even log it today (neither in perfetto nor in json).